### PR TITLE
Add unary + to AutoSIMD

### PR DESCRIPTION
**Pull request by pazner**
Add unary `operator+` to `AutoSIMD`.

This is used in the 2D templated code. I added a compile-time option for 2D to performance/ex1[p], and some sample runs/tests to detect regressions like this.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2629](https://github.com/mfem/mfem/pull/2629) | @pazner | @tzanio | @camierjs + @tzanio | 11/02/21 | 12/17/21 | 12/19/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: miniapps/performance/ex1.cpp [hunk 1]
```diff
@@ -13,6 +13,15 @@
...
 //               ex1 -m ../../data/pipe-nurbs.mesh -std  -asm -pc ho  -sc
+//               ex1_2d -m ../../data/star.mesh -perf -mf  -pc lor
+//               ex1_2d -m ../../data/star.mesh -perf -asm -pc ho
+//               ex1_2d -m ../../data/star.mesh -perf -asm -pc ho -sc
+//               ex1_2d -m ../../data/star.mesh -std  -asm -pc ho
+//               ex1_2d -m ../../data/star.mesh -std  -asm -pc ho -sc
+//               ex1_2d -m ../../data/amr-quad.mesh -perf -asm -pc ho -sc
+//               ex1_2d -m ../../data/amr-quad.mesh -std  -asm -pc ho -sc
+//               ex1_2d -m ../../data/disc-nurbs.mesh -perf -asm -pc ho  -sc
+//               ex1_2d -m ../../data/disc-nurbs.mesh -std  -asm -pc ho  -sc
```
Comment (v-dobrev) [inline miniapps/performance/ex1.cpp:24]: The scripts that extract the sample runs may need to be updated to support this. Currently, the assumption is that the executable name is the same as the source name with the suffix, `.cpp`, removed.

File: linalg/simd/auto.hpp [hunk 2]
```diff
@@ -117,6 +117,11 @@ struct MFEM_AUTOSIMD_ALIGN__ AutoSIMD
       return r;
    }
 
+   inline MFEM_ALWAYS_INLINE AutoSIMD operator+() const
+   {
+      return *this;
+   }
```
Comment (v-dobrev) [inline linalg/simd/auto.hpp:123]: I think, this method should also be added to the other files in this directory -- they add specialization for this class for specific architectures. To try this on a supported architecture, e.g. `x86`, enable `MFEM_USE_SIMD=YES`.

Comment (pazner) [reply linalg/simd/auto.hpp:123]: I added them. Is there a way to test that all of these are working properly?

Comment (v-dobrev) [reply linalg/simd/auto.hpp:123]: You can enable `MFEM_USE_SIMD=YES` and then set the architecture option in the makefile in `miniapps/performance`, e.g. for GCC here: https://github.com/mfem/mfem/blob/d09ab725b55cf86455d42b4611dccd1a835242ed/miniapps/performance/makefile#L93

For GCC and clang with `x86`, you can try replacing `-march=native` with: `-march=X` where `X` is `core2`, `skylake`, `skylake-avx512` to use respectively `m128.hpp`, `m256.hpp`, `m512.hpp`. The example prints the selected SIMD width here: https://github.com/mfem/mfem/blob/d09ab725b55cf86455d42b4611dccd1a835242ed/miniapps/performance/ex1.cpp#L127-L128

Comment (pazner) [reply linalg/simd/auto.hpp:123]: Ok, thanks. It appears to be working for the ones I was able to test.

File: miniapps/performance/ex1.cpp (hunk 1 above)
Comment (pazner) [reply miniapps/performance/ex1.cpp:24]: @v-dobrev, what do you think about something [like this](https://gist.github.com/pazner/0c09211b6b506be545e3d6fec8646b07) for including 2D and 3D in the same executable using a template?

Comment (v-dobrev) [reply miniapps/performance/ex1.cpp:24]: I think this looks good. Thanks!

Comment (pazner) [reply miniapps/performance/ex1.cpp]: Ok, great. I just pushed the changes.

**Comment by tzanio** [comment]
This looks good to me now.

@v-dobrev and @camierjs can you please review?

@pazner, can you also mention it in `CHANGELOG`?

**Review by tzanio — APPROVED** [review]

**Review by v-dobrev — APPROVED** [review]
I did not look very carefully but the issues I raised were fixed, I think. I have not tested them though, e.g. to make sure the new sample runs work fine.

**Comment by pazner** [comment]
> I did not look very carefully but the issues I raised were fixed, I think. I have not tested them though, e.g. to make sure the new sample runs work fine.

I tested the sample runs using the `config/sample-runs.sh` script, and they looked fine to me.

**Review by camierjs — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**